Add ability to hide a product from POS app

bolatovumar opened the pull request:
See discussion here: https://github.com/btcpayserver/btcpayserver/discussions/2827

| Before|After|
|---|---|
|![Screen Shot 2021-09-05 at 3 03 38 PM](https://user-images.githubusercontent.com/1934678/132145641-72b28aaa-a03c-44ba-9f54-229566e66478.png)|![Screen Shot 2021-09-05 at 3 02 46 PM](https://user-images.githubusercontent.com/1934678/132145651-aaffef91-fcab-4da5-b29f-b87a6f621bbc.png)|

bolatovumar commented:
![Screen Shot 2021-09-05 at 5 23 44 PM](https://user-images.githubusercontent.com/1934678/132145599-d0948c21-b347-4ce7-8236-a8a14f70e58b.png)

Failing test is unrelated

dstrukt commented:
Concept ACK!

@Zaxounette Any wording suggestions? 

"Mark as hidden"? "Hide Item"? To maybe be a little more specific?

dennisreimann reviewed (commented):
I think it would be better to exclude hidden items when fetching them instead of skipping them in the tempate.

Other than that it looks good to me and I think this is a good addition!

Kukks reviewed (changes requested):
The functionality is more "enabling/disabling" than hiding, so I would prefer it renamed

pavlenex commented:
> The functionality is more "enabling/disabling" than hiding, so I would prefer it renamed

Wanted to say the same.

bolatovumar commented:
> The functionality is more "enabling/disabling" than hiding, so I would prefer it renamed

@Kukks @pavlenex renamed "hidden" to "unavailable" here: 868814900b1e15f952e1001aa12b0ea7b3159187. Does this name make sense?

> I think it would be better to exclude hidden items when fetching them instead of skipping them in the tempate.
> 
> Other than that it looks good to me and I think this is a good addition!

@dennisreimann the problem here is if I exclude these items at when they are fetched it will also hide them in the template editor since it's using the same source of data as the template. I guess I could have the template editor and the templates using different data properties but it feels like a bit much (maybe I'm wrong?).

dennisreimann commented:
@bolatovumar The [`AppService.Parse` method](https://github.com/btcpayserver/btcpayserver/blob/master/BTCPayServer/Services/Apps/AppService.cs#L313) could take a parameter to configure the fetching behavior, so that it could vary depending on the use case.

Zaxounette reviewed (changes requested):
> @Kukks @pavlenex renamed "hidden" to "unavailable" here: 8688149. Does this name make sense?

I'd replace that with `Enabled <> Disabled` purely and simply. With both words visible in the toggle UI.

bolatovumar commented:
> @bolatovumar The [`AppService.Parse` method](https://github.com/btcpayserver/btcpayserver/blob/master/BTCPayServer/Services/Apps/AppService.cs#L313) could take a parameter to configure the fetching behavior, so that it could vary depending on the use case.

@dennisreimann Alright, I added a parameter to filter hidden items out just for the app FE and not the template editor in this commit: 4040cff8ee2db1ac75544f72f2475d0f1ffd97ac

Also renamed "Unavailable" to "Disabled" here: 0d9dce0025101120308012a50936766dc27f5625

dennisreimann reviewed (approved):
Code looks good. The simple parameter works for now and could easily be extended if that‘s needed in the future.

Kukks commented on BTCPayServer/Services/Apps/AppService.cs:
```diff
@@ -310,7 +311,7 @@ public string SerializeTemplate(ViewPointOfSaleViewModel.Item[] items)
             return serializer.Serialize(mappingNode);
         }
-        public ViewPointOfSaleViewModel.Item[] Parse(string template, string currency)
+        public ViewPointOfSaleViewModel.Item[] Parse(string template, string currency, bool filterDisabled = false)
```
Meh, the parser should be agnostic of business logic imo (like it was before) but it's a small thing we can change later if we want.

bolatovumar replied:
@Kukks added test here: 48c22e8db48057b0467b59da03fb8ef1ce861565

Also updated the `Parse` method not to accept `filterDisabled` and do filtering elsewhere: 40964014cd927a88c34fed7c4f379b2de038d81c

Kukks reviewed (changes requested):
Can you add a test so that if disabled, an invoice isnt created?

bolatovumar commented:
Selenium tests failing for random reasons, it doesn't really matter.

Kukks reviewed (approved):
verynice.gif
